pyteal/ast/abi/uint.py

```diff
@@ -0,0 +1,225 @@
+from typing import Union, cast
+from abc import abstractmethod
+
+from ...types import TealType
+from ...errors import TealInputError
+from ..expr import Expr
+from ..seq import Seq
+from ..assert_ import Assert
+from ..substring import Suffix
+from ..int import Int
+from ..bytes import Bytes
+from ..unaryexpr import Itob
+from ..binaryexpr import GetByte, ExtractUint16, ExtractUint32, ExtractUint64
+from ..ternaryexpr import SetByte
+from .type import Type
+
+NUM_BITS_IN_BYTE = 8
+
+
+class Uint(Type):
+    def __init__(self, bit_size: int) -> None:
+        valueType = TealType.uint64 if bit_size <= 64 else TealType.bytes
+        super().__init__(valueType)
+        self.bit_size = bit_size
+
+    def has_same_type_as(self, other: Type) -> bool:
+        return isinstance(other, Uint) and self.bit_size == other.bit_size
+
+    def is_dynamic(self) -> bool:
+        return False
+
+    def bits(self) -> int:
+        return self.bit_size
+
+    def byte_length_static(self) -> int:
+        return self.bit_size // NUM_BITS_IN_BYTE
+
+    def get(self) -> Expr:
+        return self.stored_value.load()
+
+    @abstractmethod
+    def set(self, value: Union[int, Expr]) -> Expr:
+        pass
+
+    def __str__(self) -> str:
+        return "uint{}".format(self.bit_size)
+
+
+class Uint8(Uint):
+    def __init__(
+        self,
+    ) -> None:
+        super().__init__(8)
+
+    def new_instance(self) -> "Uint8":
+        return Uint8()
+
+    def set(self, value: Union[int, Expr, "Uint8"]) -> Expr:
+        checked = False
+        if type(value) is int:
+            if value >= 2 ** 8:
+                raise TealInputError("Value exceeds Uint8 maximum: {}".format(value))
+            value = Int(value)
+            checked = True
+
+        if type(value) is Uint8:
+            value = value.get()
+            checked = True
+
+        if checked:
+            return self.stored_value.store(cast(Expr, value))
+
+        return Seq(
+            self.stored_value.store(cast(Expr, value)),
+            Assert(self.stored_value.load() < Int(2 ** self.bit_size)),
```

Comment:
The assertion is here so we fail as early as possible if the program tries to put a value that's too large into this integer type.

Previously this check was in the `encode` method, but I realized that was too late, since the value might be used in a computation before being encoded.

And the fact that the assertion happens after the store is because this order is more efficient if the value being stored is a compound expression like `Int(1) + Int(2)`. It's the difference between

```
int 1
int 2
+
int 2 ** self.bit_size # not valid code, but this is the value from pyteal
<
assert
int 1
int 2
+
store X
```

and

```
int 1
int 2
+
store X
load X
int 2 ** self.bit_size
<
assert
```

Practically asserting before or after the value is stored is the same thing from a correctness point of view.

Reply:
I see, I overlooked the perspective of generating smaller quantity of code (since it's `Expr`). That's a good point.